doc/doc-site/docs/off-chain/dss/mqtt.md

```diff
@@ -0,0 +1,77 @@
+# MQTT
+
+Econia's [DSS](./data-service-stack.md) provides an MQTT server for real-time notifications.
+It uses Mosquitto as the MQTT server, and a custom Rust program to publish PostgreSQL events to MQTT.
+
+By default, the MQTT server runs on port 21883.
+
+All messages sent on all topics are in JSON format.
+
+There are multiple available MQTT libraries across a wide variety of languages:
+
+- [mqtt.js](https://github.com/mqttjs/MQTT.js) for JavaScript/TypeScript
+- [rumqtt](https://github.com/bytebeamio/rumqtt) for Rust
+- [paho.mqtt.python](https://github.com/eclipse/paho.mqtt.python) for Python
+- many more [here](https://github.com/eclipse?q=paho.mqtt)
+
+## MQTT protocol overview
+
+Each MQTT message has a topic and a payload.
+
+You can subscribe to topics.
+
+When specifying the topic you want to subscribe to, you can put a + instead of a value to subscribe to all events matching the rest of the topic.
+
+### Example:
+
+We use the topic `fill/MARKET_ID/USER_ADDRESS/CUSTODIAN_ID`.
+Fill events by the user `0xdeadbeef` with custodian ID 1 on market 3 will have `fill/3/0xdeadbeef/1` as a topic.
+But you could subscribe to `fill/+/0xdeadbeef/+` to get all fill events from the user `0xdeadbeef`.
+
+## Topics
+
+### PlaceLimitOrderEvents
+
+`place_limit_order/MARKET_ID/USER_ADDRESS/CUSTODIAN_ID/INTEGRATOR`
+
+The JSON format for this message is the same as the REST API `/place_limit_order_events` endpoint.
+
+### PlaceMarketOrderEvents
+
+`place_market_order/MARKET_ID/USER_ADDRESS/CUSTODIAN_ID/INTEGRATOR`
+
+The JSON format for this message is the same as the REST API `/place_market_order_events` endpoint.
+
+### PlaceSwapOrderEvents
+
+`place_swap_order/MARKET_ID/INTEGRATOR/SIGNING_ACCOUNT`
+
+The JSON format for this message is the same as the REST API `/place_swap_order_events` endpoint.
+
+### ChangeOrderSizeEvents
+
+`change_order_size/MARKET_ID/USER_ADDRESS/CUSTODIAN_ID`
+
+The JSON format for this message is the same as the REST API `/change_order_size_events` endpoint.
+
+### CancelOrderEvents
+
+`cancel_order/MARKET_ID/USER_ADDRESS/CUSTODIAN_ID`
+
+The JSON format for this message is the same as the REST API `/cancel_order_events` endpoint.
+
+### FillEvents
+
+`fill/MARKET_ID/USER_ADDRESS/CUSTODIAN_ID`
+
+The JSON format for this message is the same as the REST API `/fill_events` endpoint.
+
+Note that if you subscribe to fill events for two different user/custodian ID combinations on the same market and they fill against eachother, you will recieve the same fill event notification twice, once on each channel.
+
+## Example
+
+You can find a Python example on how to use the MQTT protocol in `src/python/sdk/examples/event.py`.
+
+First, run `poetry install` to install all the necessary dependencies.
+Then, run `poetry run event` to start the example script.
```

Comment:
Presently, the `event.py` script assumes that a DSS is running against mainnet, however that will take a long time to backfill and won't be too helpful for the sake of demonstration 

For the example I think you should be able to copy-paste most of what is on our docs site currently at https://econia.dev/off-chain/dss/websocket#example , and just adapt it to use MQTT

The idea is that `event.py` by default expects to run against a local testnet DSS and can immediately show the events generated by the `trade.py` script - as was the case when we had Websockets

This will make the MQTT onboarding way easier for the newest release

Reply:
So I'm not sure if I missed a commit while reviewing this because since your comment (@alnoki) there haven't been any changes to `event.py` but it does work with the local stack as long I just update the `compose.dss-local.yaml` file. I just removed the `processor` service and included `compose.dss-core.yaml` in it and everything works

I have a few more miscellaneous thoughts on the tutorial example and such, I'll bring it up in the eng call tomorrow, but here they are before I forget:

1. Would be nice to link the REST API endpoint documentation to the JSON format section on the MQTT page for each topic type
2. We can add a sentence about running `poetry run trade` at the same time as well
3. I believe all the default values (empty input) for `trade.py` and `event.py` will just work, so we can note that as well just to make it really easy to run the example
4. Maybe repeat that short tip about rebuilding the docker images in case there are any connectivity errors
5. Provide an `local-example.env` or something for ease of use. I spent more time than I should have fighting with the silly `.env` file 😅

Reply:
I consider my contributions to this conversation resolved, standing by for @xbtmatt to mark conversation resolved